src/backend/distributed/commands/multi_copy.c

```diff
@@ -204,20 +219,101 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
 		}
 	}
 
-	partitionMethod = PartitionMethod(tableId);
-	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
+	isCopyFromWorker = IsCopyFromWorker(copyStatement);
+	if (isCopyFromWorker)
 	{
-		CopyToExistingShards(copyStatement, completionTag);
+		CopyFromWorkerNode(copyStatement, completionTag);
 	}
-	else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+	else
 	{
-		CopyToNewShards(copyStatement, completionTag);
+		char partitionMethod = PartitionMethod(tableId);
+		if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod ==
+			DISTRIBUTE_BY_RANGE)
+		{
+			CopyToExistingShards(copyStatement, completionTag);
+		}
+		else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+		{
+			CopyToNewShards(copyStatement, completionTag, NULL);
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("unsupported partition method")));
+		}
 	}
-	else
+}
+
+
+/*
+ * IsCopyFromWorker checks if the given copy statement has the master host option.
+ */
+bool
+IsCopyFromWorker(CopyStmt *copyStatement)
+{
+	ListCell *optionCell = NULL;
+	foreach(optionCell, copyStatement->options)
+	{
+		DefElem *defel = (DefElem *) lfirst(optionCell);
+		if (strncmp(defel->defname, "master_host", NAMEDATALEN) == 0)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * CopyFromWorkerNode implements the COPY table_name FROM ... from worker nodes
+ * for append-partitioned tables.
+ */
+static void
+CopyFromWorkerNode(CopyStmt *copyStatement, char *completionTag)
+{
+	NodeAddress *masterNodeAddress = MasterNodeAddress(copyStatement);
+	char *nodeName = masterNodeAddress->nodeName;
+	int32 nodePort = masterNodeAddress->nodePort;
+
+	char *nodeUser = CurrentUserName();
+	PGconn *masterConnection = ConnectToNode(nodeName, nodePort, nodeUser);
+
+	char *relationName = copyStatement->relation->relname;
+	PGresult *queryResult = NULL;
+
+	char partitionMethod = RemotePartitionMethod(relationName, masterConnection);
+	if (partitionMethod != DISTRIBUTE_BY_APPEND)
+	{
+		ereport(ERROR, (errmsg("copy from worker nodes are only supported "
+							   "for append-partitioned tables")));
+	}
+
+	/* run all metadata commands in a transaction */
+	queryResult = PQexec(masterConnection, "BEGIN");
+	if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+	{
+		ereport(ERROR, (errmsg("failed to start updating the metadata on the "
+							   "master node")));
+	}
+
+	PQclear(queryResult);
+
+	/*
+	 * Remove master node options from the copy statement becasue they are not
+	 * recognized by PostgreSQL machinery.
+	 */
+	RemoveMasterOptions(copyStatement);
+	CopyToNewShards(copyStatement, completionTag, masterConnection);
```

Comment:
> Any thoughts on making masterConnection global instead of passing it down?

If we create a global value, where should we initialize it? We can connect at the beginning of CopyFromWorkerNode(), and close connection at the end of CopyFromWorkerNode(). We can use global variable for just passing the connection.

Generally, I feel like global variables are dangerous. Do you think we really need a global variable here? What do you think?

PS. I was going to reply your previous comment :)